### ISSUE TITLE
Compilation fails on OSX 10.7

### ISSUE THREAD

**Issue by alexcrichton** [libgit2/libgit2#3386]
It looks like some of the names referenced in `stream_transport.c` aren't available in older versions of `Security.framework` perhaps?

```
bash-3.2$ cmake --build . 
[  0%] Building C object CMakeFiles/git2.dir/src/stransport_stream.c.o
In file included from /tmp/libgit2/src/stransport_stream.c:14:
In file included from /tmp/libgit2/include/git2/transport.h:10:
In file included from /tmp/libgit2/include/git2/indexer.h:10:
/tmp/libgit2/include/git2/common.h:242:11: warning: parameter '...' not found in the function declaration [-Wdocumentation]
 * @param ... value to set the option
          ^~~
/tmp/libgit2/src/stransport_stream.c:255:12: warning: implicit declaration of function 'SSLCreateContext' is invalid in C99 [-Wimplicit-function-declaration]
        st->ctx = SSLCreateContext(NULL, kSSLClientSide, kSSLStreamType);
                  ^
/tmp/libgit2/src/stransport_stream.c:255:35: error: use of undeclared identifier 'kSSLClientSide'
        st->ctx = SSLCreateContext(NULL, kSSLClientSide, kSSLStreamType);
                                         ^
/tmp/libgit2/src/stransport_stream.c:264:13: warning: implicit declaration of function 'SSLSetProtocolVersionMin' is invalid in C99 [-Wimplicit-function-declaration]
            (ret = SSLSetProtocolVersionMin(st->ctx, kTLSProtocol1)) != noErr ||
                   ^
/tmp/libgit2/src/stransport_stream.c:265:13: warning: implicit declaration of function 'SSLSetProtocolVersionMax' is invalid in C99 [-Wimplicit-function-declaration]
            (ret = SSLSetProtocolVersionMax(st->ctx, kTLSProtocol12)) != noErr ||
                   ^
/tmp/libgit2/src/stransport_stream.c:265:47: error: use of undeclared identifier 'kTLSProtocol12'; did you mean 'kTLSProtocol1'?
            (ret = SSLSetProtocolVersionMax(st->ctx, kTLSProtocol12)) != noErr ||
                                                     ^~~~~~~~~~~~~~
                                                     kTLSProtocol1
/System/Library/Frameworks/Security.framework/Headers/SecureTransport.h:88:2: note: 'kTLSProtocol1' declared here
        kTLSProtocol1,                          /* TLS 1.0 preferred, lower versions OK */
        ^
4 warnings and 2 errors generated.
make[2]: *** [CMakeFiles/git2.dir/src/stransport_stream.c.o] Error 1
make[1]: *** [CMakeFiles/git2.dir/all] Error 2
make: *** [all] Error 2
```

I checked and verified that `kTLSProtocol12` doesn't exist on a mac mini running 10.7 but it did exist on a mini running 10.8. I'm also not 100% sure if this is related to 10.7 or if we just haven't updated the mac in awhile (which is also true!)

**Comment by alexcrichton** [comment]
Er looks like I was a bit short sighted in parsing that error message, and there's some missing functions as well! According to [this site](https://developer.apple.com/library/mac/documentation/Security/Reference/secureTransportRef/#//apple_ref/c/func/SSLSetProtocolVersionMin) at least it looks like these functions are only available in 10.8+.

**Comment by carlosmn** [comment]
Indeed, those are the new functions. There are older ones which 10.7 would have available, but those are deprecated currently, and don't work on iOS.

As a quick fix, we could disable the use of SecureTransport for those older versions, but presumably you're building this to distribute prebuilt cargo binaries to newer systems as well, so that'd force OpenSSL on them as well, which is eugh.

The changes should be fairly small, so  assuming there's a way to check the OS X version from CMake, we can work around the differences.

**Comment by alexcrichton** [comment]
The quick fix may be good enough for us actually, we compile in libssh2 as well which I think still uses OpenSSL over SecureTransport anyway, so I believe we're still linking to OpenSSL regardless. I'm also totally fine passing a switch to cmake manually to disable this for now (e.g. still have it turned on by default), and that could avoid version detection maybe?

**Comment by carlosmn** [comment]
Fixed via #3389
